### PR TITLE
feat: Implement MenuItemSwitch component

### LINKED ISSUE #30105: [Feature]: Support for switch in MenuItemCheckbox.
### Library

React Components / v9 (@fluentui/react-components)

### Describe the feature that you would like added

Support for switch in MenuItemCheckbox.

Figma:https://www.figma.com/file/c7vSJ4fsMps4y9IFwMhgqL/V8-to-V9?type=design&node-id=337-66341&mode=dev
Image: 
image

### Have you discussed this feature with our team

Yes @layershifter 

### Additional context

I have raised the same concern in the Fluent UI React (Web) channel Link:https://teams.microsoft.com/l/message/19:86b094239256467da9dfa96ba0897ca2@thread.skype/1701683894991?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=ffe264f2-14d0-48b5-9384-64f808b81294&parentMessageId=1701683894991&teamName=Fluent%20Community&channelName=Fluent%20UI%20React%20(Web)&createdTime=1701683894991. In this same post Oleksandr Fediashov(V9 team member)  shared a POC link with me to verify whether it is feasible or not for my temporary use case. He suggested that I raise a feature request on the v9 team, as this is a valid feature request.

### Validations

- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Priority

High

### PR DESCRIPTION
**Pull request by ling1726**
menuitemswitch

Fixes #30105

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Avatar-mount_base.html">636</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Avatar-mount.html">656</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Button-mount_base.html">316</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Button-mount.html">311</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Field-mount_base.html">1155</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Field-mount.html">1108</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProvider-mount_base.html">717</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProvider-mount.html">720</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-mount_base.html">84</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-mount.html">87</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">36</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">36</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">77</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/MakeStyles-mount_base.html">881</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/MakeStyles-mount.html">893</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Persona-mount_base.html">1832</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/Persona-mount.html">1779</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/SpinButton-mount_base.html">1393</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/SpinButton-mount.html">1451</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/SwatchPicker-mount_base.html">1576</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31099/perf-test-react-components/SwatchPicker-mount.html">1602</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-menu/stories/Menu/MenuSwitchItem.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  </Menu>
+);
+
+SwitchItem.parameters = {
+  docs: {
+    description: {
+      story: [
+        'A variant of `MenuItemCheckbox` that displays selection using a switch design.',
+        "This is commonly used for menus that dont't really have strong selection function but needs",
+        'to support and exceptional selected option.',
```
Comment (spmonahan) [inline packages/react-components/react-menu/stories/Menu/MenuSwitchItem.stories.tsx:32]:
```suggestion
        'to support an exceptional selected option.',
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import type { ForwardRefComponent } from '@fluentui/react-utilities';
+import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
+import { useMenuItemSwitch_unstable } from './useMenuItemSwitch';
+import { renderMenuItemSwitch_unstable } from './renderMenuItemSwitch';
+import { useMenuItemSwitchStyles_unstable } from './useMenuItemSwitchStyles.styles';
+import type { MenuItemSwitchProps } from './MenuItemSwitch.types';
+
+/**
+ * MenuItemSwitch component - TODO: add more docs
+ */
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.tsx:11]:
```suggestion
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.test.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      name: 'checkbox',
+      value: '1',
+    },
+    displayName: 'MenuItemSwitch',
+    testOptions: {
+      'has-static-classnames': [
+        {
+          props: {
+            icon: 'Test Icon',
+            checkmark: 'Test Checkmark',
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.test.tsx:25]:
```suggestion
            switchIndicator: 'Test Switch',
```

?

File: packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.types.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+export type MenuItemSwitchSlots = Pick<MenuItemSlots, 'root' | 'content' | 'secondaryContent' | 'icon'> & {
+  switchIndicator?: Slot<'span'>;
+};
+
+/**
+ * MenuItemSwitch Props
+ */
+export type MenuItemSwitchProps = ComponentProps<MenuItemSwitchSlots> &
+  Pick<MenuItemCheckboxProps, 'disabled' | 'persistOnClick' | 'name' | 'value'> & {};
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.types.ts:14]:
```suggestion
  Pick<MenuItemCheckboxProps, 'disabled' | 'persistOnClick' | 'name' | 'value'>;
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.types.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+import { MenuItemSlots } from '../MenuItem/MenuItem.types';
+import { MenuItemSelectableState } from '../../selectable/types';
+import { MenuItemCheckboxProps } from '../MenuItemCheckbox/MenuItemCheckbox.types';
+
+export type MenuItemSwitchSlots = Pick<MenuItemSlots, 'root' | 'content' | 'secondaryContent' | 'icon'> & {
+  switchIndicator?: Slot<'span'>;
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.types.ts:7]:
```suggestion
  switch?: Slot<'span'>;
```

Just `switch`? `MenuItemCheckbox` has just `checkmark`

File: packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.test.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            submenuIndicator: 'Test Submenu Indicator',
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/MenuItemSwitch.test.tsx:26]:
```suggestion
```

?

File: packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+// Thumb and track sizes used by the component.
+const spaceBetweenThumbAndTrack = 2;
+const trackHeight = 20;
+const trackWidth = 40;
+const thumbSize = trackHeight - spaceBetweenThumbAndTrack;
+
+const useSwitchIndicatorBaseClassName = makeResetStyles({
+  borderRadius: tokens.borderRadiusCircular,
+  border: `${tokens.strokeWidthThin} solid`,
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts:25]:
```suggestion
  border: `${tokens.strokeWidthThin} solid ${tokens.colorNeutralStrokeAccessible}`,
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  transitionProperty: 'background, border, color',
+  width: `${trackWidth}px`,
+  marginRight: tokens.spacingHorizontalXS,
+
+  '@media screen and (prefers-reduced-motion: reduce)': {
+    transitionDuration: '0.01ms',
+  },
+
+  color: tokens.colorNeutralStrokeAccessible,
+  ...shorthands.borderColor(tokens.colorNeutralStrokeAccessible),
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts:44]:
```suggestion
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  ':hover': {
+    color: tokens.colorNeutralStrokeAccessibleHover,
+    ...shorthands.borderColor(tokens.colorNeutralStrokeAccessibleHover),
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitchStyles.styles.ts:47]:
```suggestion
    borderColor: tokens.colorNeutralStrokeAccessibleHover,
```

File: packages/react-components/react-menu/stories/Menu/MenuSwitchItem.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    </MenuPopover>
+  </Menu>
+);
+
+SwitchItem.parameters = {
+  docs: {
+    description: {
+      story: [
+        'A variant of `MenuItemCheckbox` that displays selection using a switch design.',
+        "This is commonly used for menus that dont't really have strong selection function but needs",
```
Comment (layershifter) [inline packages/react-components/react-menu/stories/Menu/MenuSwitchItem.stories.tsx:30]:
```suggestion
        "This is commonly used for menus that don't really have strong selection function but needs",
```

File: packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitch.tsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  ref: React.Ref<HTMLDivElement>,
+): MenuItemSwitchState => {
+  const baseState = useMenuItemCheckbox_unstable(props, ref);
+  return {
+    ...baseState,
+    switchIndicator: slot.optional(props.switchIndicator, {
+      renderByDefault: true,
+      elementType: 'span',
+      defaultProps: {
+        children: <CircleFilled className={circleFilledClassName} />,
```
Comment (layershifter) [inline packages/react-components/react-menu/src/components/MenuItemSwitch/useMenuItemSwitch.tsx:28]: In my PoC I made a custom `SwitchBox` component to be used as indicator: https://stackblitz.com/edit/stackblitz-starters-qjhegq?file=src%2FSwitchBox.tsx

I don't have strong opinions there, but styles where a bit cleaner there as nested selectors are not needed

**Review by spmonahan — APPROVED** [review]

**Pull request merged**